File: src/Inverse.java
import java.math.BigInteger;

public class Inverse extends ExtendedEuclidean {
    private BigInteger N, x; // Zn Group Number: N, Number: x
    private BigInteger xInv, xAnotherInv; // Inverse of x
    private boolean isExist = false, nGreaterThanX = false;

    public Inverse(BigInteger N, BigInteger x) {
        super(N, x);

        this.N = N;
        this.x = x;

        if(N.compareTo(x) <= 0) { // N < x
            nGreaterThanX = true;
            xInv = null;
            xAnotherInv = null;
        } else if(getGcd().compareTo(BigInteger.ONE) == 0) { // For inverse to exist, gcd must be 1
            isExist = true;
            if(getB().compareTo(BigInteger.ZERO) < 0) { // inverse is negative
                xInv = getB();
                xAnotherInv = getB().add(getX());
            } else { // inverse is positive
                xInv = getB();
                xAnotherInv = getB().subtract(getX());
            }
        } else {
            xInv = null;
            xAnotherInv = null;
        }
    }

    public String toString() {
        if(isExist) return "Inverse of " + x.toString() + " is " + xInv.toString() + "(or " + xAnotherInv.toString() + ") in Z " + N.toString() + " Group";
        else {
            if(nGreaterThanX) return "N must be greater than x";
            else return x.toString() + " doesn't have inverse in Z " + N.toString() + " Group";
        }
    }
}

File: src/python/Inverse.py
'''
    Inverse
     - param:  (a, N)
        a: Number to find inverse
        N: Number of Z group
     - return: inverse a
'''

class Inverse:
    def __init__(self):
        self.inv = 0

    # 1. Euclidean Algorithm
    def gcd(self, a, b):
        if a < b:
            a, b = b, a

        if b == 0:
            return a
        # if a % b == 0:
        #     return b
        else:
            return self.gcd(b, a%b)

    # 2. Extended Euclidean Algorithm
    def eea(self, a, b):
        if a < b:
            a, b = b, a
        
        r0, r1 = a, b
        s0, s1 = 1, 0
        t0, t1 = 0, 1

        while r1 > 0:
            q = r0 // r1
            r = r0 - q*r1
            r0, r1 = r1, r

            s = s0 - q*s1
            s0, s1 = s1, s

            t = t0 - q*t1
            t0, t1 = t1, t
        
        self.inv = t0
        return "gcd({}, {}) = ({})*{} + ({})*{} = {}".format(a, b, s0, a, t0, b, r0)

    # 3. Inverse
    def inverse(self, N, a):
        if N < a:
            return "N must be greater than a"
        
        if self.gcd(N,a) == 1:
            self.eea(N, a)
            # if self.inv < 0:
            #     return "{} (or {})".format(N+self.inv, self.inv)
            # else:
            #     return "{} (or {})".format(self.inv, self.inv-N)
            return self.inv
        else:
            return "{} doesn't have inverse in Z {} Group".format(a, N)




##########
## Test ##
##########
inv = Inverse()

# print(inv.gcd(18, 12))
# print(inv.gcd(53, 17))

# print(inv.eea(18, 12))
# print(inv.eea(120, 44))

# print(inv.inverse(12, 18))
# print(inv.inverse(18, 12))
# print(inv.inverse(53, 17))
# print(inv.inverse(100, 23))